MSVC internal compiler error building 1.9.0 `test/utils.c`

```
[218/219] Compiling C object subprojects/libxkbcommon-xkbcommon-1.9.0/test-utils.exe.p/test_utils.c.obj
FAILED: subprojects/libxkbcommon-xkbcommon-1.9.0/test-utils.exe.p/test_utils.c.obj 
"cl" "-Isubprojects\libxkbcommon-xkbcommon-1.9.0\test-utils.exe.p" "-Isubprojects\libxkbcommon-xkbcommon-1.9.0" "-I..\subprojects\libxkbcommon-xkbcommon-1.9.0" "-I..\subprojects\libxkbcommon-xkbcommon-1.9.0\src" "-I..\subprojects\libxkbcommon-xkbcommon-1.9.0\include" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W4" "/std:c11" "/Od" "/Zi" "/utf-8" "/wd4100" "/Fdsubprojects\libxkbcommon-xkbcommon-1.9.0\test-utils.exe.p\test_utils.c.pdb" /Fosubprojects/libxkbcommon-xkbcommon-1.9.0/test-utils.exe.p/test_utils.c.obj "/c" ../subprojects/libxkbcommon-xkbcommon-1.9.0/test/utils.c
../subprojects/libxkbcommon-xkbcommon-1.9.0/test/utils.c(266): warning C4034: sizeof returns 0
../subprojects/libxkbcommon-xkbcommon-1.9.0/test/utils.c(271): warning C4034: sizeof returns 0
../subprojects/libxkbcommon-xkbcommon-1.9.0/test/utils.c(289): warning C4034: sizeof returns 0
../subprojects/libxkbcommon-xkbcommon-1.9.0/test/utils.c(398): warning C4244: 'function': conversion from 'time_t' to 'unsigned int', possible loss of data
D:\a\wrapdb\wrapdb\subprojects\libxkbcommon-xkbcommon-1.9.0\test\utils.c(153) : fatal error C1001: Internal compiler error.
(compiler file 'D:\a\_work\1\s\src\vctools\Compiler\Utc\src\p2\main.cpp', line 255)
 To work around this problem, try simplifying or changing the program near the locations listed above.
If possible please provide a repro here: https://developercommunity.visualstudio.com/ 
Please choose the Technical Support command on the Visual C++ 
 Help menu, or open the Technical Support help file for more information
  cl!InvokeCompilerPassW()+0x1b433
  cl!InvokeCompilerPassW()+0x1b433
  cl!InvokeCompilerPassW()+0x1b400
  cl!InvokeCompilerPassW()+0x1adbc
  cl!InvokeCompilerPassW()+0x1ad3b
  cl!InvokeCompilerPassW()+0x1afd7
  cl!InvokeCompilerPassW()+0x1af0e
```

Full setup log:
```
The Meson build system
Version: 1.8.0.rc2
Source dir: D:\a\wrapdb\wrapdb
Build dir: D:\a\wrapdb\wrapdb\_build
Build type: native build
Project name: wrapdb
Project version: undefined
Host machine cpu family: x86_64
libxkbcommon| Configuring xkeyboard-config-test using configuration
libxkbcommon| Configuring test-config.h using configuration
libxkbcommon| Library m found: NO
libxkbcommon| Program python3 (jinja2) found: NO
libxkbcommon| Program test/symbols-leak-test.py found: YES (C:\hostedtoolcache\windows\Python\3.9.13\x64\python.exe D:\a\wrapdb\wrapdb\subprojects\libxkbcommon-xkbcommon-1.9.0\test/symbols-leak-test.py)
libxkbcommon| Program valgrind found: NO
libxkbcommon| Message: valgrind not found, disabling valgrind test setup
libxkbcommon| Header "getopt.h" has symbol "getopt_long" : NO
libxkbcommon| Configuring config.h using configuration
libxkbcommon| Build targets in project: 30
libxkbcommon| Subproject libxkbcommon finished.

Dependency xkbcommon found: YES 1.9.0 (overridden)
Build targets in project: 31

libxkbcommon 1.9.0

  Compiler
    backend              : ninja
    buildtype            : debug
    id                   : msvc
    c_args               :
    c_link_args          :
    yacc                 : C:\ProgramData\Chocolatey\bin\win_bison.EXE 2.7

  Directories
    prefix               : c:/
    bindir               : bin
    libdir               : lib
    datadir              : share
    xkb-config-root      : c:/share/X11/xkb
    xkb-config-extra-path: c:/etc/xkb
    xlocaledir           : c:/share/X11/locale

  Features
    docs                 : false
    tools                : false
    wayland              : false
    x11                  : false

  Defaults
    layout               : us
    model                : pc105
    options              : 
    rules                : evdev
    variant              : 

  Tests
    merge-modes          : false

wrapdb undefined

  Subprojects
    libxkbcommon                   : YES

  User defined options
    depnames                       : xkbcommon
    libxkbcommon:enable-docs       : false
    libxkbcommon:enable-tools      : false
    libxkbcommon:enable-wayland    : false
    libxkbcommon:enable-x11        : false
    libxkbcommon:enable-xkbregistry: false
    prognames                      : 
    python.install_env             : auto
    wraps                          : libxkbcommon

Found ninja-1.12.1 at C:\ProgramData\Chocolatey\bin\ninja.EXE
```

CI runs [here](https://github.com/mesonbuild/wrapdb/actions/runs/14687388340/job/41217838688?pr=2068) and [here](https://github.com/mesonbuild/wrapdb/actions/runs/14687388340/job/41217838690?pr=2068).

--- a/src/xkbcomp/symbols.c
+++ b/src/xkbcomp/symbols.c
@@ -1024,8 +1024,11 @@ SetSymbolsField(SymbolsInfo *info, KeyInfo *keyi, const char *field,
         }
         else {
             ndx--;
-            if (ndx >= darray_size(keyi->groups))
+            if (ndx >= darray_size(keyi->groups)) {
+                /* Avoid clang-tidy false positive */
+                assert(darray_size(keyi->groups) < ndx + 1);
                 darray_resize0(keyi->groups, ndx + 1);
+            }
             darray_item(keyi->groups, ndx).type = val;
             darray_item(keyi->groups, ndx).defined |= GROUP_FIELD_TYPE;
         }

--- a/test/atom.c
+++ b/test/atom.c
@@ -59,9 +59,6 @@ test_random_strings(void)
     table = atom_table_new();
     assert(table);
 
-    unsigned seed = (unsigned) clock();
-    srand(seed);
-
     N = 1 + rand() % 100000;
     arr = calloc(N, sizeof(*arr));
     assert(arr);
@@ -81,7 +78,6 @@ test_random_strings(void)
                         strlen(string), string);
                 fprintf(stderr, "new length %zu, string %.*s\n",
                         arr[i].len, (int) arr[i].len, arr[i].string);
-                fprintf(stderr, "seed: %u\n", seed);
                 assert(false);
             }
 
@@ -95,7 +91,6 @@ test_random_strings(void)
         if (arr[i].atom == XKB_ATOM_NONE) {
             fprintf(stderr, "failed to intern! len: %zu, string: %.*s\n",
                     arr[i].len, (int) arr[i].len, arr[i].string);
-            fprintf(stderr, "seed: %u\n", seed);
             assert(false);
         }
     }
@@ -121,7 +116,6 @@ test_random_strings(void)
             }
             fprintf(stderr, "END\n");
 
-            fprintf(stderr, "seed: %u\n", seed);
             assert(false);
         }
     }
@@ -132,13 +126,28 @@ test_random_strings(void)
     atom_table_free(table);
 }
 
+/* CLI positional arguments:
+ * 1. Seed for the pseudo-random generator:
+ *    - Leave it unset or set it to “-” to use current time.
+ *    - Use an integer to set it explicitly.
+ */
 int
-main(void)
+main(int argc, char *argv[])
 {
     struct atom_table *table;
     xkb_atom_t atom1, atom2, atom3;
 
     test_init();
+
+    /* Initialize pseudo-random generator with program arg or current time */
+    unsigned int seed;
+    if (argc >= 2 && !streq(argv[1], "-")) {
+        seed = (unsigned int) atoi(argv[1]);
+    } else {
+        seed = (unsigned int) time(NULL);
+    }
+    fprintf(stderr, "Seed for the pseudo-random generator: %u\n", seed);
+    srand(seed);
 
     table = atom_table_new();
     assert(table);

--- a/test/compose.c
+++ b/test/compose.c
@@ -4,6 +4,7 @@
  */
 
 #include "config.h"
+#include <stdlib.h>
 #include <time.h>
 #include <errno.h>
 #include <stdio.h>
@@ -1066,13 +1067,13 @@ main(int argc, char *argv[])
     assert(ctx);
 
     /* Initialize pseudo-random generator with program arg or current time */
-    int seed;
+    unsigned int seed;
     if (argc >= 2 && !streq(argv[1], "-")) {
-        seed = atoi(argv[1]);
+        seed = (unsigned int) atoi(argv[1]);
     } else {
-        seed = (int)time(NULL);
+        seed = (unsigned int) time(NULL);
     }
-    fprintf(stderr, "Seed for the pseudo-random generator: %d\n", seed);
+    fprintf(stderr, "Seed for the pseudo-random generator: %u\n", seed);
     srand(seed);
 
     /* Determine number of loops for quickchecks */
@@ -1119,5 +1120,5 @@ main(int argc, char *argv[])
     test_roundtrip(ctx);
 
     xkb_context_unref(ctx);
-    return 0;
+    return EXIT_SUCCESS;
 }

--- a/test/utils.c
+++ b/test/utils.c
@@ -154,7 +154,7 @@ test_number_parsers(void)
     /* Check the claim that it always works on normal strings using SIZE_MAX and
      * that it always stops on the first NULL byte */
     {
-        const struct {
+        static const struct {
             const char* input;
             struct { int count; uint64_t val; } dec;
             struct { int count; uint64_t val; } hex;
@@ -170,9 +170,39 @@ test_number_parsers(void)
                 .hex = { 0, UINT64_C(0) }
             },
             {
+                .input = "/",
+                .dec = { 0, UINT64_C(0) },
+                .hex = { 0, UINT64_C(0) }
+            },
+            {
+                .input = ";",
+                .dec = { 0, UINT64_C(0) },
+                .hex = { 0, UINT64_C(0) }
+            },
+            {
                 .input = "x",
                 .dec = { 0, UINT64_C(0) },
                 .hex = { 0, UINT64_C(0) }
+            },
+            {
+                .input = "/1",
+                .dec = { 0, UINT64_C(0) },
+                .hex = { 0, UINT64_C(0) }
+            },
+            {
+                .input = ";1",
+                .dec = { 0, UINT64_C(0) },
+                .hex = { 0, UINT64_C(0) }
+            },
+            {
+                .input = "x1",
+                .dec = { 0, UINT64_C(0) },
+                .hex = { 0, UINT64_C(0) }
+            },
+            {
+                .input = "0",
+                .dec = { 1, UINT64_C(0) },
+                .hex = { 1, UINT64_C(0) }
             },
             {
                 .input = "1",
@@ -200,14 +230,44 @@ test_number_parsers(void)
                 .hex = { -1, UINT64_C(0x1844674407370955)}
             },
             {
+                .input = "18446744073709551616",
+                .dec = { -1, UINT64_C(1844674407370955161)},
+                .hex = { -1, UINT64_C(0x1844674407370955) }
+            },
+            {
+                .input = "99999999999999999999",
+                .dec = { -1, UINT64_C(9999999999999999999)},
+                .hex = { -1, UINT64_C(0x9999999999999999) }
+            },
+            {
                 .input = "184467440737095516150",
                 .dec = { -1, UINT64_MAX                  },
                 .hex = { -1, UINT64_C(0x1844674407370955)}
             },
             {
+                .input = "00000000000000000",
+                .dec = { 17, UINT64_C(0) },
+                .hex = { 17, UINT64_C(0) }
+            },
+            {
+                .input = "00000000000000001",
+                .dec = { 17, UINT64_C(1) },
+                .hex = { 17, UINT64_C(1) }
+            },
+            {
                 .input = "ffffffffffffffff",
                 .dec = { 0 , 0         },
                 .hex = { 16, UINT64_MAX}
+            },
+            {
+                .input = "ffffffffffffffff0",
+                .dec = { 0 , 0         },
+                .hex = { -1, UINT64_MAX}
+            },
+            {
+                .input = "10000000000000000",
+                .dec = { 17, UINT64_C(10000000000000000) },
+                .hex = { -1, UINT64_C(0x1000000000000000)}
             },
             {
                 .input = "fffffffffffffffff",
@@ -217,91 +277,62 @@ test_number_parsers(void)
 
         };
         for (size_t k = 0; k < ARRAY_SIZE(tests); k++) {
-            uint64_t dec = 0;
-            int count;
-            count = parse_dec_to_uint64_t(tests[k].input, SIZE_MAX, &dec);
-            assert_printf(count == tests[k].dec.count,
-                          "SIZE_MAX #%zu, expected: %d, got: %d\n",
-                          k, tests[k].dec.count, count);
-            assert_printf(dec == tests[k].dec.val,
-                          "SIZE_MAX #%zu, expected: %"PRIu64", got: %"PRIu64"\n",
-                          k, tests[k].dec.val, dec);
-            uint64_t hex = 0;
-            count = parse_hex_to_uint64_t(tests[k].input, SIZE_MAX, &hex);
-            assert_printf(count == tests[k].hex.count,
-                          "SIZE_MAX #%zu, expected: %d, got: %d\n",
-                          k, tests[k].hex.count, count);
-            assert_printf(hex == tests[k].hex.val,
-                          "SIZE_MAX #%zu, expected: %#"PRIx64", got: %#"PRIx64"\n",
-                          k, tests[k].hex.val, hex);
+            const size_t len = strlen(tests[k].input);
+            /* Try different lengths */
+            const struct {const char* label; size_t len;} sizes[] = {
+                { .label = "buffer"  , .len = len     },
+                { .label = "string"  , .len = len + 1 },
+                { .label = "SIZE_MAX", .len = SIZE_MAX}
+            };
+            for (unsigned int s = 0; s < ARRAY_SIZE(sizes); s++) {
+                int count;
+                /* Decimal */
+                uint64_t dec = 0;
+                count =
+                    parse_dec_to_uint64_t(tests[k].input, sizes[s].len, &dec);
+                assert_printf(count == tests[k].dec.count,
+                              "Dec %s #%zu \"%s\" (%zu), "
+                              "expected: %d, got: %d\n",
+                              sizes[s].label, k, tests[k].input, sizes[s].len,
+                              tests[k].dec.count, count);
+                assert_printf(dec == tests[k].dec.val,
+                              "Dec %s #%zu \"%s\", "
+                              "expected: %"PRIu64", got: %"PRIu64"\n",
+                              sizes[s].label, k, tests[k].input,
+                              tests[k].dec.val, dec);
+                /* Hexadecimal */
+                uint64_t hex = 0;
+                count =
+                    parse_hex_to_uint64_t(tests[k].input, sizes[s].len, &hex);
+                assert_printf(count == tests[k].hex.count,
+                              "Hex %s #%zu \"%s\" (%zu), "
+                              "expected: %d, got: %d\n",
+                              sizes[s].label, k, tests[k].input, sizes[s].len,
+                              tests[k].hex.count, count);
+                assert_printf(hex == tests[k].hex.val,
+                              "Hex %s #%zu \"%s\", "
+                              "expected: %#"PRIx64", got: %#"PRIx64"\n",
+                              sizes[s].label, k, tests[k].input,
+                              tests[k].hex.val, hex);
+            }
         }
     }
 
-#define str_safe_len(input) \
-    (sizeof(input) == 0 ?   \
-        0                   \
-        : sizeof(input) - ((input)[sizeof(input) - 1] == '\0' ? -1 : 0))
 #define PRIuint64_t PRIx64
 #define PRIuint32_t PRIx32
-#define test_parse_to(type, format, input, count, expected) do {             \
-    type n = 0;                                                              \
-    int r = parse_##format##_to_##type(input, str_safe_len(input), &n);      \
-    assert_printf(r == (count),                                              \
-                  "expected count: %d, got: %d (value: %#"PRI##type", string: %s)\n", \
-                  count, r, n, input);                                       \
-    assert_printf(n == (expected),                                           \
-                  "expected value: %#"PRI##type", got: %#"PRI##type"\n", expected, n);         \
+#define test_parse_to(type, format, input, count, expected) do {     \
+    type n = 0;                                                      \
+    int r = parse_##format##_to_##type(input, ARRAY_SIZE(input), &n);\
+    assert_printf(r == (count),                                      \
+                  "Buffer: expected count: %d, "                     \
+                  "got: %d (value: %#"PRI##type", string: %.*s)\n",  \
+                  count, r, n, (int) ARRAY_SIZE(input), (input));    \
+    assert_printf(n == (expected),                                   \
+                  "Buffer: expected value: %#"PRI##type", got: "     \
+                  "%#"PRI##type"\n", expected, n);                   \
 } while (0)
 
-    char empty[] = {};
-    char not_null_terminated_0[] = { '0' };
-    char not_null_terminated_1[] = { '1' };
-    char not_null_terminated_dec_max[] = {
-        '1', '8', '4', '4', '6', '7', '4', '4', '0', '7', '3', '7', '0', '9',
-        '5', '5', '1', '6', '1', '5'
-    };
-    char buffer[30] = {0};
-    int count;
-
-    test_parse_to(uint64_t, dec, empty, 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, not_null_terminated_0, 1, UINT64_C(0));
-    test_parse_to(uint64_t, dec, not_null_terminated_1, 1, UINT64_C(1));
-    test_parse_to(uint64_t, dec, not_null_terminated_dec_max,
-                  (int) sizeof(not_null_terminated_dec_max), UINT64_MAX);
-    test_parse_to(uint64_t, dec, empty, 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "/", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, ";", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "x", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "/0", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, ";0", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "x0", 0, UINT64_C(0));
-    test_parse_to(uint64_t, dec, "18446744073709551616", -1, UINT64_MAX / 10);
-    test_parse_to(uint64_t, dec, "184467440737095516150", -1, UINT64_MAX);
-    test_parse_to(uint64_t, dec, "99999999999999999999", -1,
-                  UINT64_C(9999999999999999999));
-
-    char not_null_terminated_hex_max[] = {
-        'f', 'f', 'f', 'f', 'f', 'f', 'f', 'f',
-        'f', 'f', 'f', 'f', 'f', 'f', 'f', 'f',
-    };
-
-    test_parse_to(uint64_t, hex, empty, 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, not_null_terminated_0, 1, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, not_null_terminated_1, 1, UINT64_C(0x1));
-    test_parse_to(uint64_t, hex, not_null_terminated_hex_max,
-                  (int) sizeof(not_null_terminated_hex_max), UINT64_MAX);
-    test_parse_to(uint64_t, hex, "", 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, "/", 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, ";", 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, "x", 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, "xf", 0, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, "00000000000000000", 17, UINT64_C(0x0));
-    test_parse_to(uint64_t, hex, "00000000000000001", 17, UINT64_C(0x1));
-    test_parse_to(uint64_t, hex, "ffffffffffffffff0", -1, UINT64_MAX);
-    test_parse_to(uint64_t, hex, "10000000000000000", -1,
-                  UINT64_C(0x1000000000000000));
-
+    /* Test syntax variants */
     const uint64_t values[] = {
         UINT64_C(0),
         UINT64_C(1),
@@ -324,7 +355,9 @@ test_number_parsers(void)
         UINT64_MAX - 1,
         UINT64_MAX,
     };
+    char buffer[30] = {0};
     for (size_t k = 0; k < ARRAY_SIZE(values); k++) {
+        int count;
         /* Basic: decimal */
         count = snprintf(buffer, sizeof(buffer), "%"PRIu32, (uint32_t) values[k]);
         assert(count > 0);
@@ -399,6 +432,7 @@ test_number_parsers(void)
     for (unsigned k = 0; k < 10000; k++) {
         const uint32_t x32 = rand_uint32();
         const uint64_t x64 = rand_uint64();
+        int count;
         /* Hex: Lower case */
         count = snprintf(buffer, sizeof(buffer), "%"PRIx32, x32);
         assert(count > 0);

--- a/test/utils.c
+++ b/test/utils.c
@@ -102,12 +102,6 @@ test_path_functions(void)
 static uint32_t
 rand_uint32(void)
 {
-    static int initialized = 0;
-    if (!initialized) {
-        srand((unsigned int)time(NULL));
-        initialized = 1;
-    }
-
     /* First decide how many bits we’ll actually use (1-32) */
     int bits = 1 + (rand() % 32);
 
@@ -126,12 +120,6 @@ rand_uint32(void)
 static uint64_t
 rand_uint64(void)
 {
-    static int initialized = 0;
-    if (!initialized) {
-        srand((unsigned int)time(NULL));
-        initialized = 1;
-    }
-
     /* First decide how many bits we’ll actually use (1-64) */
     int bits = 1 + (rand() % 64);
 
@@ -428,7 +416,6 @@ test_number_parsers(void)
     }
 
     /* Random */
-    srand(time(NULL));
     for (unsigned k = 0; k < 10000; k++) {
         const uint32_t x32 = rand_uint32();
         const uint64_t x64 = rand_uint64();
@@ -471,10 +458,26 @@ test_number_parsers(void)
 }
 /* NOLINTEND(google-readability-function-size) */
 
+/* CLI positional arguments:
+ * 1. Seed for the pseudo-random generator:
+ *    - Leave it unset or set it to “-” to use current time.
+ *    - Use an integer to set it explicitly.
+ */
 int
-main(void)
+main(int argc, char *argv[])
 {
     test_init();
+
+    /* Initialize pseudo-random generator with program arg or current time */
+    unsigned int seed;
+    if (argc >= 2 && !streq(argv[1], "-")) {
+        seed = (unsigned int) atoi(argv[1]);
+    } else {
+        seed = (unsigned int) time(NULL);
+    }
+    fprintf(stderr, "Seed for the pseudo-random generator: %u\n", seed);
+    srand(seed);
+
     test_string_functions();
     test_path_functions();
     test_number_parsers();

